include/aspect/particle/generator/interface.h

```diff
@@ -0,0 +1,177 @@
+#define __aspect__particle_generator_interface_h
+
+#include <aspect/particle/particle.h>
+#include <aspect/plugins.h>
+#include <aspect/simulator_access.h>
+
+#include <deal.II/base/parameter_handler.h>
+
+namespace aspect
+{
+  namespace Particle
+  {
+    /**
+     * A namespace in which we define everything that has to do with defining
+     * the particle generation.
+     *
+     * @ingroup ParticleGenerators
+     */
+    namespace Generator
+    {
+      /**
+       * Abstract base class used for classes that generate particles.
+       *
+       * @ingroup ParticleGenerators
+       */
+      template <int dim>
+      class Interface : public SimulatorAccess<dim>
+      {
+        public:
+          /**
+           * Constructor.
+           */
+          Interface() {}
+
+          /**
+           * Destructor. Made virtual so that derived classes can be created
+           * and destroyed through pointers to the base class.
+           */
+          virtual ~Interface () {}
+
+          /**
+           * Generate particles. Every derived class
+           * has to decide on the method and number of particles to generate,
+           * for example using input parameters declared in their
+           * declare_parameters and parse_parameters functions. This function
+           * should generate the particles and associate them to their according
+           * cells by inserting them into a multimap between cell and particle.
+           *
+           * @return A multimap containing cells and their contained particles.
+           */
+          virtual
+          std::multimap<types::LevelInd, Particle<dim> >
+          generate_particles() = 0;
+
+
+          /**
+           * Declare the parameters this class takes through input files. The
+           * default implementation of this function does not describe any
+           * parameters. Consequently, derived classes do not have to overload
+           * this function if they do not take any runtime parameters.
+           */
+          static
+          void
+          declare_parameters (ParameterHandler &prm);
+
+          /**
+           * Read the parameters this class declares from the parameter file.
+           * The default implementation of this function does not read any
+           * parameters. Consequently, derived classes do not have to overload
+           * this function if they do not take any runtime parameters.
+           */
+          virtual
+          void
+          parse_parameters (ParameterHandler &prm);
+
+        protected:
+          /**
+           * Generate a particle at the specified position and with the
+           * specified id. Many derived classes use this functionality,
+           * therefore it is implemented here to avoid duplication.
+           */
+          std::pair<types::LevelInd,Particle<dim> >
+          generate_particle(const Point<dim> &position,
+                            const types::particle_index id) const;
```

Comment:
By the way, could this function be made `static`, or does it actually access any member functions?